Add columnMappings to explain plan output

zachjsh opened the pull request:
This change adds `columnMappings` to the native explain query mode explain plan output. This field is a list containing the mapping of queryColumn name to outputColumn name. 

As an example for this rollup query (as given as an [example](https://druid.apache.org/docs/latest/multi-stage-query/examples.html#insert-with-rollup) in the MSQ docs)

```
INSERT INTO "kttm_rollup"

WITH kttm_data AS (
SELECT * FROM TABLE(
  EXTERN(
    '{"type":"http","uris":["https://static.imply.io/example-data/kttm-v2/kttm-v2-2019-08-25.json.gz"]}',
    '{"type":"json"}',
    '[{"name":"timestamp","type":"string"},{"name":"agent_category","type":"string"},{"name":"agent_type","type":"string"},{"name":"browser","type":"string"},{"name":"browser_version","type":"string"},{"name":"city","type":"string"},{"name":"continent","type":"string"},{"name":"country","type":"string"},{"name":"version","type":"string"},{"name":"event_type","type":"string"},{"name":"event_subtype","type":"string"},{"name":"loaded_image","type":"string"},{"name":"adblock_list","type":"string"},{"name":"forwarded_for","type":"string"},{"name":"language","type":"string"},{"name":"number","type":"long"},{"name":"os","type":"string"},{"name":"path","type":"string"},{"name":"platform","type":"string"},{"name":"referrer","type":"string"},{"name":"referrer_host","type":"string"},{"name":"region","type":"string"},{"name":"remote_address","type":"string"},{"name":"screen","type":"string"},{"name":"session","type":"string"},{"name":"session_length","type":"long"},{"name":"timezone","type":"string"},{"name":"timezone_offset","type":"long"},{"name":"window","type":"string"}]'
  )
))

SELECT
  FLOOR(TIME_PARSE("timestamp") TO MINUTE) AS __time,
  session,
  agent_category,
  agent_type,
  browser,
  browser_version,
  MV_TO_ARRAY("language") AS "language", -- Multi-value string dimension
  os,
  city,
  country,
  forwarded_for AS ip_address,

  COUNT(*) AS "cnt",
  SUM(session_length) AS session_length,
  APPROX_COUNT_DISTINCT_DS_HLL(event_type) AS unique_event_types
FROM kttm_data
WHERE os = 'iOS'
GROUP BY 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11
PARTITIONED BY HOUR
CLUSTERED BY browser, session
```

The plan with added `columnMappings` field is given as:

```
[
  {
    "query": {
      "queryType": "groupBy",
      "dataSource": {
        "type": "external",
        "inputSource": {
          "type": "http",
          "uris": [
            "https://static.imply.io/example-data/kttm-v2/kttm-v2-2019-08-25.json.gz"
          ]
        },
        "inputFormat": {
          "type": "json",
          "keepNullColumns": false,
          "assumeNewlineDelimited": false,
          "useJsonNodeReader": false
        },
        "signature": [
          {
            "name": "timestamp",
            "type": "STRING"
          },
          {
            "name": "agent_category",
            "type": "STRING"
          },
          {
            "name": "agent_type",
            "type": "STRING"
          },
          {
            "name": "browser",
            "type": "STRING"
          },
          {
            "name": "browser_version",
            "type": "STRING"
          },
          {
            "name": "city",
            "type": "STRING"
          },
          {
            "name": "continent",
            "type": "STRING"
          },
          {
            "name": "country",
            "type": "STRING"
          },
          {
            "name": "version",
            "type": "STRING"
          },
          {
            "name": "event_type",
            "type": "STRING"
          },
          {
            "name": "event_subtype",
            "type": "STRING"
          },
          {
            "name": "loaded_image",
            "type": "STRING"
          },
          {
            "name": "adblock_list",
            "type": "STRING"
          },
          {
            "name": "forwarded_for",
            "type": "STRING"
          },
          {
            "name": "language",
            "type": "STRING"
          },
          {
            "name": "number",
            "type": "LONG"
          },
          {
            "name": "os",
            "type": "STRING"
          },
          {
            "name": "path",
            "type": "STRING"
          },
          {
            "name": "platform",
            "type": "STRING"
          },
          {
            "name": "referrer",
            "type": "STRING"
          },
          {
            "name": "referrer_host",
            "type": "STRING"
          },
          {
            "name": "region",
            "type": "STRING"
          },
          {
            "name": "remote_address",
            "type": "STRING"
          },
          {
            "name": "screen",
            "type": "STRING"
          },
          {
            "name": "session",
            "type": "STRING"
          },
          {
            "name": "session_length",
            "type": "LONG"
          },
          {
            "name": "timezone",
            "type": "STRING"
          },
          {
            "name": "timezone_offset",
            "type": "LONG"
          },
          {
            "name": "window",
            "type": "STRING"
          }
        ]
      },
      "intervals": {
        "type": "intervals",
        "intervals": [
          "-146136543-09-08T08:23:32.096Z/146140482-04-24T15:36:27.903Z"
        ]
      },
      "virtualColumns": [
        {
          "type": "expression",
          "name": "v0",
          "expression": "timestamp_floor(timestamp_parse(\"timestamp\",null,'UTC'),'PT1M',null,'UTC')",
          "outputType": "LONG"
        },
        {
          "type": "expression",
          "name": "v1",
          "expression": "mv_to_array(\"language\")",
          "outputType": "ARRAY<STRING>"
        },
        {
          "type": "expression",
          "name": "v2",
          "expression": "'iOS'",
          "outputType": "STRING"
        }
      ],
      "filter": {
        "type": "selector",
        "dimension": "os",
        "value": "iOS"
      },
      "granularity": {
        "type": "all"
      },
      "dimensions": [
        {
          "type": "default",
          "dimension": "v0",
          "outputName": "d0",
          "outputType": "LONG"
        },
        {
          "type": "default",
          "dimension": "session",
          "outputName": "d1",
          "outputType": "STRING"
        },
        {
          "type": "default",
          "dimension": "agent_category",
          "outputName": "d2",
          "outputType": "STRING"
        },
        {
          "type": "default",
          "dimension": "agent_type",
          "outputName": "d3",
          "outputType": "STRING"
        },
        {
          "type": "default",
          "dimension": "browser",
          "outputName": "d4",
          "outputType": "STRING"
        },
        {
          "type": "default",
          "dimension": "browser_version",
          "outputName": "d5",
          "outputType": "STRING"
        },
        {
          "type": "default",
          "dimension": "v1",
          "outputName": "d6",
          "outputType": "ARRAY<STRING>"
        },
        {
          "type": "default",
          "dimension": "v2",
          "outputName": "d7",
          "outputType": "STRING"
        },
        {
          "type": "default",
          "dimension": "city",
          "outputName": "d8",
          "outputType": "STRING"
        },
        {
          "type": "default",
          "dimension": "country",
          "outputName": "d9",
          "outputType": "STRING"
        },
        {
          "type": "default",
          "dimension": "forwarded_for",
          "outputName": "d10",
          "outputType": "STRING"
        }
      ],
      "aggregations": [
        {
          "type": "count",
          "name": "a0"
        },
        {
          "type": "longSum",
          "name": "a1",
          "fieldName": "session_length"
        },
        {
          "type": "HLLSketchBuild",
          "name": "a2",
          "fieldName": "event_type",
          "lgK": 12,
          "tgtHllType": "HLL_4",
          "round": true
        }
      ],
      "limitSpec": {
        "type": "default",
        "columns": [
          {
            "dimension": "d4",
            "direction": "ascending",
            "dimensionOrder": {
              "type": "lexicographic"
            }
          },
          {
            "dimension": "d1",
            "direction": "ascending",
            "dimensionOrder": {
              "type": "lexicographic"
            }
          }
        ]
      },
      "context": {
        "finalizeAggregations": false,
        "groupByEnableMultiValueUnnesting": false,
        "queryId": "665106f9-325a-4799-85a8-cc16855bca10",
        "sqlInsertSegmentGranularity": "\"HOUR\"",
        "sqlQueryId": "665106f9-325a-4799-85a8-cc16855bca10"
      }
    },
    "signature": [
      {
        "name": "d0",
        "type": "LONG"
      },
      {
        "name": "d1",
        "type": "STRING"
      },
      {
        "name": "d2",
        "type": "STRING"
      },
      {
        "name": "d3",
        "type": "STRING"
      },
      {
        "name": "d4",
        "type": "STRING"
      },
      {
        "name": "d5",
        "type": "STRING"
      },
      {
        "name": "d6",
        "type": "ARRAY<STRING>"
      },
      {
        "name": "d7",
        "type": "STRING"
      },
      {
        "name": "d8",
        "type": "STRING"
      },
      {
        "name": "d9",
        "type": "STRING"
      },
      {
        "name": "d10",
        "type": "STRING"
      },
      {
        "name": "a0",
        "type": "LONG"
      },
      {
        "name": "a1",
        "type": "LONG"
      },
      {
        "name": "a2",
        "type": "LONG"
      }
    ],
    "columnMappings": [
      {
        "queryColumn": "d0",
        "outputColumn": "__time"
      },
      {
        "queryColumn": "d1",
        "outputColumn": "session"
      },
      {
        "queryColumn": "d2",
        "outputColumn": "agent_category"
      },
      {
        "queryColumn": "d3",
        "outputColumn": "agent_type"
      },
      {
        "queryColumn": "d4",
        "outputColumn": "browser"
      },
      {
        "queryColumn": "d5",
        "outputColumn": "browser_version"
      },
      {
        "queryColumn": "d6",
        "outputColumn": "language"
      },
      {
        "queryColumn": "d7",
        "outputColumn": "os"
      },
      {
        "queryColumn": "d8",
        "outputColumn": "city"
      },
      {
        "queryColumn": "d9",
        "outputColumn": "country"
      },
      {
        "queryColumn": "d10",
        "outputColumn": "ip_address"
      },
      {
        "queryColumn": "a0",
        "outputColumn": "cnt"
      },
      {
        "queryColumn": "a1",
        "outputColumn": "session_length"
      },
      {
        "queryColumn": "a2",
        "outputColumn": "unique_event_types"
      }
    ]
  }
]

```

This PR has:

- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

abhishekrb19 commented on sql/src/main/java/org/apache/druid/sql/calcite/planner/QueryUtils.java:
```diff
@@ -0,0 +1,58 @@
+
+  /**
+   * Builds the mappings for queryColumn to outputColumn
+   * @param fieldMapping The field mappigns
```
```suggestion
   * @param fieldMapping The field mappings
```

zachjsh replied:
fixed

abhishekrb19 commented on sql/src/main/java/org/apache/druid/sql/calcite/planner/QueryUtils.java:
```diff
@@ -0,0 +1,58 @@
+   * @param druidQuery The Druid query
+   * @return Mappings for queryColumn to outputColumn
+   */
+  public static List<ColumnMapping> buildColumnMappings(
```
It'd be great if we can add a few UTs for this public method.

zachjsh replied:
talked offline, and decided UTs here wouldnt be very useful as this function is very simple, and already fully covered in other tests.

abhishekrb19 reviewed (commented):
LGTM overall. A few questions and comments:
1. Will the new `columnMappings` be backwards compatible with existing clients? Currently, this info is added as part of `PLAN` column along with `queries` and `signature` -- I think this should be ok as it's a new map entry, but would like to double check.
2. Can we update the `EXPLAIN PLAN` sample output in the following `README` to include the new info: https://github.com/apache/druid/blob/master/docs/querying/sql-translation.md#interpreting-explain-plan-output?
3. Since `columnMappings` info is only available in the native explain query mode (and not in the legacy Calcite mode),  it'd be worth clarifying that in the PR summary or title, so it's evident in the next release notes.

gianm commented on .github/scripts/unit_tests_script.sh:
```diff
@@ -62,8 +62,8 @@ then
   --type jacoco \
-  --line-coverage 50 \
-  --branch-coverage 50 \
+  --line-coverage 1 \
```
Please do revert this back to 50, 50.

zachjsh replied:
yeah was intentional to get around the code coverage blocking the ITs from running, I will revert back for sure.

abhishekrb19 replied:
If this change is to get around the coverage failure and run everything, I believe you could achieve that by running workflows directly from the source repo's branch, i.e., your fork

gianm reviewed (commented):
The basic idea LGTM. I generally agree with @abhishekrb19's review comments.

zachjsh commented on sql/src/main/java/org/apache/druid/sql/calcite/planner/QueryHandler.java:
```diff
@@ -355,6 +355,7 @@ public void cleanup(QueryHandler.EnumeratorIterator<Object[]> iterFromMake)
    * Construct a {@link PlannerResult} for an 'explain' query from a {@link RelNode}
    */
   protected PlannerResult planExplanation(
+      final RelRoot relRoot,
```
fixed

zachjsh commented on sql/src/main/java/org/apache/druid/sql/calcite/planner/QueryUtils.java:
```diff
@@ -0,0 +1,58 @@
+    final List<ColumnMapping> columnMappings = new ArrayList<>();
+    for (final Pair<Integer, String> entry : fieldMapping) {
+      final String queryColumn = druidQuery.getOutputRowSignature().getColumnName(entry.getKey());
+      final String outputColumns = entry.getValue();
```
fixed

zachjsh commented:
> LGTM overall. A few questions and comments:
> 
>     1. Will the new `columnMappings` be backwards compatible with existing clients? Currently, this info is added as part of `PLAN` column along with `queries` and `signature` -- I think this should be ok as it's a new map entry, but would like to double check.
> 
>     2. Can we update the `EXPLAIN PLAN` sample output in the following `README` to include the new info: https://github.com/apache/druid/blob/master/docs/querying/sql-translation.md#interpreting-explain-plan-output?
> 
>     3. Since `columnMappings` info is only available in the native explain query mode (and not in the legacy Calcite mode),  it'd be worth clarifying that in the PR summary or title, so it's evident in the next release notes.

thanks @abhishekrb19 , I addressed your concerns here. Adding a new property to existing returned object should not break existing clients, unless they are bad clients 👿

abhishekrb19 reviewed (approved):
LGTM, thanks!

zachjsh commented:
dont merge until I revert the code coverage threshold. Will do so when all tests pass.
